rc/lsp.kak

```diff
@@ -113,40 +115,180 @@
+            # as needed from diagnostics.
+            if (extra_dynamic_lines < 0) {
+                # If the diagnostics has not already been maked to be truncated
+                # we mark it to be truncated, add a static line for the
+                # "Diagnostics truncated" message, update the
+                # available_dynamic_lines and extra_dynamic_lines.
+                if (diagnostics_truncated == 0) {
+                    diagnostics_truncated = 1
+                    static_lines++
+
+                    available_dynamic_lines = available_lines > static_lines ?
+                        available_lines - static_lines : 0
+                    extra_dynamic_lines = requested_dynamic_lines > available_dynamic_lines ?
+                        requested_dynamic_lines - available_dynamic_lines : 0
+                }
+
+                diagnostic_lines = diagnostic_lines > extra_dynamic_lines ?
+                    diagnostic_lines - extra_dynamic_lines : 0
+
+                requested_dynamic_lines = info_lines + diagnostic_lines
+                extra_dynamic_lines = requested_dynamic_lines > available_dynamic_lines ?
+                    requested_dynamic_lines - available_dynamic_lines : 0
+            }
         }
-        if (code_lenses) {
-            r = r "Code Lenses available (shortcut l)\n"
-            lines++
+
+        # Track if there is anything in the hover, so we know when to insert
+        # a new-line character.
+        output_in_hover = 0
+
+        if (info_lines) {
+            printf info_data_line[1]
+            for (i = 2; i <= info_lines; i++) {
+                printf "%s%s", "\n", info_data_line[i]
+            }
+            output_in_hover = 1
         }
-        if (ENVIRON["kak_opt_lsp_modeline_code_actions"]) {
-            r = r "Code Actions available (shortcut a)\n"
-            lines++
+
+        if (info_truncated == 1) {
+            if (output_in_hover == 1) printf "\n"
+            printf "{+i@InfoDefault}Hover info truncated, use lsp-hover-buffer (shortcut H) for full hover info"
+            output_in_hover = 1
         }
 
-        if (ENVIRON["kak_opt_lsp_modeline_message_requests"]) {
-            r = r "There are unread messages (use lsp-show-message-request-next to read)\n"
-            lines++
+        if (diagnostic_data) {
+            if (output_in_hover == 1) printf "\n"
+            printf "{+b@InfoDefault}Diagnostics{InfoDefault} (shortcut e):"
+
+            if (diagnostic_lines) {
+                printf "\n"
+                printf diagnostic_data_line[1]
+                for (i = 2; i <= diagnostic_lines; i++) {
+                    printf "%s%s", "\n", diagnostic_data_line[i]
+                }
+            }
+
+            if (diagnostics_truncated == 1) {
+                printf "\n"
+                printf "{+i@InfoDefault}Hover diagnostics truncated, use lsp-hover-buffer (shortcut H) for full hover info"
```

Comment:
not sure if it's a good idea to show two messages about truncation.
I'd expect only one at the bottom. 
At the bottom, it's right next to the other hints about code actions etc. that show which shortcuts to use.